crates/holochain_state/src/nonce.rs

```diff
@@ -0,0 +1,171 @@
+use crate::mutations;
+use holochain_sqlite::nonce::get_nonce;
+use holochain_sqlite::nonce::IntNonce;
+use holochain_sqlite::prelude::DatabaseResult;
+use holochain_sqlite::prelude::DbWrite;
+use holochain_sqlite::rusqlite::named_params;
+use holochain_sqlite::sql::sql_conductor;
+use holochain_types::prelude::AgentPubKey;
+use holochain_types::prelude::DbKindConductor;
+use holochain_zome_types::Timestamp;
+use rand::Rng;
+use std::time::Duration;
+
+/// Rather arbitrary but we expire nonces after 5 mins.
+pub const FRESH_NONCE_EXPIRES_AFTER: Duration = Duration::from_secs(60 * 5);
+pub const WITNESSABLE_EXPIRY_DURATION: Duration = Duration::from_secs(600 * 5);
+
+#[derive(PartialEq, Debug)]
+pub enum WitnessNonceResult {
+    Fresh,
+    Duplicate,
+    Expired,
+    Future,
+}
+
+pub async fn witness_nonce(
+    db: &DbWrite<DbKindConductor>,
+    agent: AgentPubKey,
+    nonce: IntNonce,
+    now: Timestamp,
+    expires: Timestamp,
+) -> DatabaseResult<WitnessNonceResult> {
+    // Treat expired but also very far future expiries as stale as we cannot trust the time in that case.
+    if expires <= now {
+        Ok(WitnessNonceResult::Expired)
+    } else if expires > (now + WITNESSABLE_EXPIRY_DURATION)? {
+        Ok(WitnessNonceResult::Future)
+    } else {
+        db.async_commit(move |txn| {
+            txn.execute(
+                sql_conductor::DELETE_EXPIRED_NONCE,
+                named_params! {":now": now},
+            )?;
+            if get_nonce(txn, &agent, nonce, now)?.is_some() {
+                Ok(WitnessNonceResult::Duplicate)
+            } else {
+                mutations::insert_nonce(txn, &agent, nonce, expires)?;
+                Ok(WitnessNonceResult::Fresh)
+            }
+        })
+        .await
+    }
+}
+
+pub fn fresh_nonce(now: Timestamp) -> DatabaseResult<(IntNonce, Timestamp)> {
+    // very unlikely to get a collision, we assume impossible.
+    let nonce = rand::thread_rng().gen();
+    let expires: Timestamp = (now + FRESH_NONCE_EXPIRES_AFTER)?;
```

Comment:
I feel like we've discussed this before in relation to monotonically increasing chain timestamps, but consider [something like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=999fabf9b8ea3d2f929db0166951c7ef).

- By having the timestamp and the nonce be the same value, we no longer have to keep each nonce in a database, we can just track a single used nonce and require all future calls for that entity to have a greater nonce.
- Is there a need for callers to set varying expire values? If not, there'd be less grief opportunity if we always expect the nonce to be within a certain statically defined time window.

Reply:
Oh, is this that part where you were looking into ensuring that messages were delivered over the network in order, and we couldn't guarantee that?

In that case, we'd still need to store all nonces in the database until the predefined expiry, but we could still merge the values as far as what is signed / passed into calls : )

Reply:
@neonphog yes we have code like that for the source chain but i don't think it's appropriate here

if a user misconfigures their system clock to be far future _ever_ then they send out a timestamp-nonce with a signature on it, then they can no longer send subsequent zome calls unless they either keep the wrong system time, or track which zome call times they have sent to whom

it makes more sense to me for recipients to witness nonces than for senders to track nonces they have sent

we can change the expiry time to an issued at time with a fixed timeout sure

Reply:
@thedavidmeister - what do you think about the idea of still merging the timestamp / nonce fields but dropping the must be greater than requirement? So we still have recipients witness nonces, and multiple out-of-order messages are still valid, but we only have to send one bit of information.

Reply:
so we discussed this on a call and the summary is that we came a full circle back to the current implementation because if nothing else this system has to be implemented by _every_ client that wants to do a zome call, and we can probably expect everyone to know how to generate a random number, but maybe monotonic timestamp tracking is too subtle/stateful to impose on _every_ client

Reply:
After realtime discussion, we agreed to keep the code as-is to make it easier for clients to make calls into the system. i.e. just generate the random number.